### PR TITLE
Custom sdp implementations dense

### PR DESCRIPTION
**Pull request by drisspg**
# Summary 

- This code creates the runtime dispatch system for choosing a performant fused SDP kernel. The only choice of fused kernel is flash_attention. It also creates python flags and a context manager that can be used to turn off and on behavior for dispatch. 
- This also adds support for flash_attention with dense tensors.

### CODE REVIEW

**Comment by drisspg** [comment]
### Performance overhead of runtime disptach:

#### With Constraint checking: select_sdp_backend()
<torch.utils.benchmark.utils.common.Measurement object at 0x7f290e727790>
Flash MHA Dense
  Median: 1.33 ms
  IQR:    0.01 ms (1.33 to 1.34)
  76 measurements, 100 runs per measurement, 48 threads

#### Without Constraint Checking:
Flash MHA Dense
  Median: 1.30 ms
  IQR:    0.01 ms (1.30 to 1.31)
  77 measurements, 100 runs per measurement, 48 threads

**Comment by drisspg** [comment]
### Examlple Python MHA Class
class FlashMHA(torch.nn.Module):
    def __init__(self, num_heads, qkv, proj):
        super().__init__()
        self.qkv = qkv
        self.proj = proj
        self.num_heads = num_heads

    @sdp_kernel(enable_math=False)
    def forward(self, query: torch.Tensor, key: torch.Tensor, value: torch.Tensor):
        assert query is key
        assert key is value

        query_projected = torch.nn.functional.linear(
            query, self.qkv.weight, self.qkv.bias)

        batch_size = query_projected.size(0)
        seq_len = query_projected.size(1)
        embed_dim = query_projected.size(2)
        head_dim = embed_dim//(self.num_heads*3)

        # Transpose seq_len and num_heads dim
        query_projected = query_projected.view(batch_size, seq_len, 3*self.num_heads, head_dim).transpose(1, 2)
        query, key, value = query_projected.chunk(3, 1)

        # the output of sdp = (batch, num_heads, seq_len, head_dim)
        attn, _ = torch.nn.functional._scaled_dot_product_attention(
            query, key, value, attn_mask=None, dropout_p=0.0, need_attn_weights=False, is_causal=False)

        # Determine which one of these is faster
        attn = attn.transpose(1, 2).view(batch_size, seq_len, self.num_heads*head_dim)
        # attn = attn.transpose(1, 2).contiguous().view(batch_size, seq_len, self.num_heads*head_dim)
        attn = torch.nn.functional.linear(
            attn, self.proj.weight, self.proj.bias)
        return attn, None

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -13163,11 +13163,23 @@
...
   python_module: nn
   variants: function
   autogen: _scaled_dot_product_attention.out
 
+# Register the math kernel for cpu
+- func: _scaled_dot_product_attention_forward(Tensor query, Tensor key, Tensor value, Tensor? attn_mask=None, float dropout_p=0.0, bool need_attn_weights=False, bool is_causal=False) -> (Tensor, Tensor)
+  variants: function
+  dispatch:
+    CUDA: _scaled_dot_product_attention_forward_cuda
+    CPU: _scaled_dot_product_attention_forward_math
```
Comment (drisspg) [inline aten/src/ATen/native/native_functions.yaml:13176]: `_scaled_dot_product_attention_forward_math` is the same as _scaled_dot_product_attention_math. Since _scaled_dot_product_attention jumps straight to forward if it's inputs don't requires grad. We could add more checks at this layer but this would bubble up some logic out of the forward function which I think is undesirable.

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -13163,11 +13163,23 @@
...
+    NestedTensorCPU, NestedTensorCUDA: _scaled_dot_product_attention_forward_math
```
Comment (drisspg) [inline aten/src/ATen/native/native_functions.yaml:13177]: Since we have SDP tests as well in nested tensor we register the math kernel to forward.

File: aten/src/ATen/native/transformers/cuda/sdp_utils.h [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+  bool has_attn_mask;
+  double dropout;
+  bool need_attn_weights;
+  bool is_causal;
+};
+
+enum class SDPBackend {flash_attention, math, error};
+
+//  Define gate functions that determine if a flash kernel can be ran
+inline bool check_gradients(sdp_params params, bool debug) {
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/sdp_utils.h:32]: We already check for this at the SDP top level so this should be safe to remove, I want to be as restrictive as possible initially but this should be removed

File: aten/src/ATen/native/transformers/cuda/sdp_utils.h [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+        " gpu.");
+    return false;
+  }
+  return true;
+}
+
+inline bool use_flash_attention(sdp_params params, bool debug) {
+  // Constraints specific to flash attention
+  static const std::unordered_set<int64_t> flash_embed_dim_sizes{
+      16, 32, 64, 128};
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/sdp_utils.h:192]: I think with the refactor all head_dims%8 ==0 are supported but let me check and update if thats true

File: aten/src/ATen/native/transformers/cuda/sdp_utils.h (hunk 3 above)
Comment (cpuhrsch) [inline aten/src/ATen/native/transformers/cuda/sdp_utils.h:32]: This function seems unused

File: aten/src/ATen/native/transformers/cuda/sdp_utils.h [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        params.key.requires_grad(),
+        " Value: ",
+        params.value.requires_grad());
+    return false;
+  }
+  return true;
+}
+
+inline bool compiled_with_flash_kernels(sdp_params params, bool debug) {
+#ifndef USE_FLASH_ATTENTION
```
Comment (cpuhrsch) [inline aten/src/ATen/native/transformers/cuda/sdp_utils.h:49]: I think it makes more sense to guard `return False` within `use_flash_attention`. That'll give the compiler the chance to just entirely ignore the function instead of running a for loop on std::functions.

File: aten/src/ATen/native/transformers/cuda/sdp_utils.h [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+  TORCH_CHECK(debug, "Torch was not compiled with flash attention");
+  return false;
+#endif
+  return true;
+}
+
+inline bool check_tensor_dtype(
+    sdp_params params,
+    std::vector<caffe2::ScalarType> allowed_dtypes,
```
Comment (cpuhrsch) [inline aten/src/ATen/native/transformers/cuda/sdp_utils.h:58]: Why not make these template arguments or use one of the AT_DISPATCH macros we have?

File: aten/src/ATen/native/transformers/cuda/sdp_utils.h (hunk 3 above)
Comment (cpuhrsch) [reply aten/src/ATen/native/transformers/cuda/sdp_utils.h:32]: You can also use https://github.com/pytorch/pytorch/blob/04108592a362848a5d3af4332f7628a14e312174/c10/core/GradMode.h

**Review by cpuhrsch — APPROVED** [review]
